File: lifeos/reviews/LIFEOS-P3-088/evidence/independent_static_runner.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';

const root = process.argv[2];
const outputAt = process.argv[3] === '--output' ? process.argv[4] : undefined;
if (!root) throw new Error('Usage: node independent_static_runner.mjs <task-local-app-dir>');
const read = (name) => fs.readFileSync(path.join(root, name), 'utf8');
const files = Object.fromEntries(['default-recovery.html', 'no-reliable-suggestion.html', 'restricted-offline.html', 'app.js', 'styles.css'].map((name) => [name, read(name)]));
const checks = [];
const check = (id, condition, detail) => checks.push({ id, status: condition ? 'PASS' : 'FAIL', detail });
const pages = ['default-recovery.html', 'no-reliable-suggestion.html', 'restricted-offline.html'];

for (const page of pages) {
  const html = files[page];
  check(`${page}:language`, /<html lang="zh-CN">/i.test(html), 'declares Chinese document language');
  check(`${page}:viewport`, /name="viewport" content="width=device-width, initial-scale=1"/.test(html), 'declares responsive viewport');
  check(`${page}:skip-link`, /class="skip-link" href="#main-content"/.test(html), 'provides keyboard skip link');
  check(`${page}:main-landmark`, /<main id="main-content"/.test(html), 'has main landmark matching skip link');
  check(`${page}:navigation`, /<nav class="page-nav" aria-label="今日状态导航">/.test(html), 'labels state navigation');
  check(`${page}:three-local-pages`, /href="default-recovery.html"/.test(html) && /href="no-reliable-suggestion.html"/.test(html) && /href="restricted-offline.html"/.test(html), 'only relative local page navigation');
  check(`${page}:boundary-copy`, /AI 未启用/.test(html) && /网络未使用/.test(html), 'states AI and network closed');
  check(`${page}:local-assets`, /href="styles.css"/.test(html) && /src="app.js"/.test(html), 'uses only local CSS and JS');
  check(`${page}:no-remote-url`, !/(?:https?:|\/\/)/i.test(html), 'contains no remote URL');
}

const defaultPage = files['default-recovery.html'];
check('default:semantic-heading-order', /<h1>今日<\/h1>[\s\S]*<h2 id="project-title">[\s\S]*<h3>/.test(defaultPage), 'uses h1 then h2 then h3 in main recovery content');
check('default:original-label', /你的记录／原文/.test(defaultPage), 'labels user-origin text');
check('default:explicit-confirmation', /确认保留在本次会话/.test(defaultPage), 'requires explicit confirmation');
check('default:live-status', /id="capture-status" role="status" aria-live="polite"/.test(defaultPage), 'announces interaction state');
check('default:no-autosave-attribute', !/autofocus|oninput|localStorage|sessionStorage/i.test(defaultPage), 'does not imply automatic or persistent capture');

const emptyPage = files['no-reliable-suggestion.html'];
check('empty:no-fabricated-suggestion', /暂无可靠建议/.test(emptyPage) && /不会虚构重点或建议/.test(emptyPage), 'fails closed when evidence is insufficient');
check('empty:two-controlled-paths', /id="choose-project"/.test(emptyPage) && /id="record-stop"/.test(emptyPage), 'offers two manual controlled paths');
check('empty:live-status', /id="no-suggestion-status" role="status" aria-live="polite"/.test(emptyPage), 'announces chosen path');

const restrictedPage = files['restricted-offline.html'];
check('restricted:fail-closed-copy', /不读取来源、不处理内容，也不生成建议/.test(restrictedPage), 'states restricted/offline fail-closed behavior');
check('restricted:no-network-sync', /不联网、不同步/.test(restrictedPage), 'states network and sync closed');

const js = files['app.js'];
check('script:empty-input-rejected', /if \(!value\) return clear\('未显示记录：请先输入原文，再明确确认。'\)/.test(js), 'rejects empty input');
check('script:confirmation-is-explicit', /confirm-save/.test(js) && /addEventListener\('click'/.test(js), 'confirmation is click-triggered');
check('script:failure-clears-record', /simulate-failure[\s\S]*clear\('模拟失败：本次会话未保留或显示原文/.test(js), 'simulated failure clears displayed record');
check('script:no-storage-or-network', !/(localStorage|sessionStorage|indexedDB|fetch\(|XMLHttpRequest|WebSocket|navigator\.sendBeacon|FileReader|showOpenFilePicker|tauri|invoke\()/i.test(js), 'does not access storage, network, file, or Tauri APIs');
check('script:manual-empty-state-paths', /选择 Project 是受控路径/.test(js) && /先记录当前停点是受控路径/.test(js), 'two empty-state paths do not read material');

const css = files['styles.css'];
check('css:visible-focus', /:focus-visible[^{]*\{[^}]*outline:3px solid var\(--focus\)/.test(css), 'uses high-contrast visible focus outline');
check('css:skip-link-focus', /\.skip-link:focus \{ top:12px; \}/.test(css), 'reveals skip link on focus');
check('css:narrow-layout', /@media \(max-width:600px\)[\s\S]*grid-template-columns:1fr/.test(css), 'switches action and navigation layout on narrow screens');
check('css:readable-container', /width:min\(100% - 40px, 900px\)/.test(css), 'uses constrained readable layout');
check('css:reduced-motion', /prefers-reduced-motion:reduce/.test(css), 'honors reduced-motion preference');
check('css:no-remote-import', !/@import|url\(/i.test(css), 'contains no remote stylesheet or asset import');

for (const [name, content] of Object.entries(files)) {
  check(`closed:${name}`, !/(?:https?:|wss?:|localStorage|sessionStorage|indexedDB|fetch\(|XMLHttpRequest|WebSocket|FileReader|showOpenFilePicker|tauri|ipc|supabase|firebase|openai)/i.test(content), 'does not contain prohibited real-capability token');
}

const result = {
  runner: 'LIFEOS-P3-088 independent static runner',
  target: root,
  generated_at: new Date().toISOString(),
  counts: { pass: checks.filter((item) => item.status === 'PASS').length, fail: checks.filter((item) => item.status === 'FAIL').length },
  hashes: Object.fromEntries(Object.entries(files).map(([name, content]) => [name, crypto.createHash('sha256').update(content).digest('hex')])),
  checks,
};
const serialized = JSON.stringify(result, null, 2);
if (outputAt) fs.writeFileSync(outputAt, `${serialized}\n`);
console.log(serialized);
process.exitCode = result.counts.fail ? 1 : 0;

File: lifeos/reviews/LIFEOS-P3-090/rework/attempt-2/evidence/independent_static_runner.py
#!/usr/bin/env python3
"""P3-090 attempt-2 independent, read-only static review runner."""
import hashlib
import json
import sys
from pathlib import Path

APP = Path(sys.argv[1]).resolve()
OUT = Path(sys.argv[2]).resolve()
ROOT = Path(__file__).resolve().parents[6]

expected = {
    "default-recovery.html": "c8e6e130bdb26962f6663afc56529fecf2c3501f643633655837daa433ee16b4",
    "no-reliable-suggestion.html": "c38a3d8fe35bb00d5d13ca1fa1967daa813c1b8af083e75de3fb9cb593d0520d",
    "restricted-offline.html": "0965726629e105087b94680476f18e5a35f361f046be23ccbc468850ac84085f",
    "app.js": "88dc10ffa945d12baf42b771e4c6fddef791a5f1305e9f1d349161977684fe1a",
    "styles.css": "052600b12672a7d9bf255c130edbb4eb5d6d3d3bfee3a64be403e24e74dfe94a",
}
history = {
    "P3-079 integrated_runtime.py": (ROOT / "lifeos/engineering/LIFEOS-P3-079/src/integrated_runtime.py", "19d337a560cfa3e6098571b29a71fc121e059db1c0914e3e3f58f3f740905788"),
    "P3-087 app.js": (ROOT / "lifeos/engineering/LIFEOS-P3-087/app.js", "fa50eb0e7bfa177a3ba2c6e97ec1185370506e84cc2f1d01434ac9d5ea090d0c"),
    "P3-088 independent runner": (ROOT / "lifeos/reviews/LIFEOS-P3-088/evidence/independent_static_runner.mjs", "ec2966b3e92c759482444dc25d6e83eb8243b21b27541a870928f4186cd2cb8e"),
}
def sha(path): return hashlib.sha256(path.read_bytes()).hexdigest()
results = []
def check(case, condition, detail): results.append({"id": case, "pass": bool(condition), "detail": detail})

for name, digest in expected.items():
    check("HASH-" + name, (APP / name).exists() and sha(APP / name) == digest, "task-local source hash matches P3-089 manifest")
for name, (path, digest) in history.items():
    check("HIST-" + name, path.exists() and sha(path) == digest, "specified read-only historical hash unchanged")

html = {name: (APP / name).read_text(encoding="utf-8") for name in expected if name.endswith(".html")}
js = (APP / "app.js").read_text(encoding="utf-8")
css = (APP / "styles.css").read_text(encoding="utf-8")
all_text = "\n".join(html.values()) + "\n" + js + "\n" + css
for name, page in html.items():
    check("PAGE-" + name + "-local", 'href="styles.css"' in page and 'src="app.js"' in page, "only relative local CSS/JS resources")
    check("PAGE-" + name + "-skip", 'class="skip-link" href="#main-content"' in page, "skip link targets main content")
    check("PAGE-" + name + "-boundary", "AI 未启用" in page and ("网络未使用" in page or "网络未使用；" in page), "synthetic and AI/network boundary is visible")
    check("PAGE-" + name + "-nav", page.count(".html") >= 3 and 'aria-label="今日状态导航"' in page, "three-page navigation is present")
check("LIFE-default-deny", "默认拒绝" in html["default-recovery.html"] and "默认拒绝" in js, "default-deny is explicit in UI and state")
check("LIFE-capture-confirm", "confirm-capture" in html["default-recovery.html"] and "明确确认" in js, "capture needs explicit confirmation")
check("LIFE-grant-revoke", "id=\"grant\"" in html["default-recovery.html"] and "id=\"revoke\"" in html["default-recovery.html"] and "明确 grant" in js, "grant and revoke controls are distinct")
check("LIFE-confirm-exact", "restore-confirmation" in html["default-recovery.html"] and "!== 'CONFIRM'" in js, "restore requires exact CONFIRM")
check("LIFE-failure-clear", "simulate-failure" in html["default-recovery.html"] and "clearDisplay(" in js and "失败披露" in js, "failure clears display and discloses failure")
check("LIFE-session-only", "当前页面 DOM" in all_text and "刷新或关闭后状态清除" in all_text, "session-only disclosure is explicit")
check("STATE-no-suggestion", "不读取资料、不生成建议" in html["no-reliable-suggestion.html"] and "暂无可靠建议" in html["no-reliable-suggestion.html"], "no fabricated suggestion")
check("STATE-restricted", "权限受限 · 离线" in html["restricted-offline.html"] and "已阻断" in html["restricted-offline.html"], "restricted/offline remains fail-closed")
check("A11Y-focus", ":focus-visible" in css and "outline:3px" in css, "visible focus styling exists")
check("A11Y-responsive", "@media (max-width:600px)" in css and "grid-template-columns:1fr" in css, "narrow-screen layout exists")
check("A11Y-reduced-motion", "prefers-reduced-motion:reduce" in css, "reduced-motion override exists")
for forbidden in ["http://", "https://", "fetch(", "XMLHttpRequest", "localStorage", "sessionStorage", "indexedDB", "navigator.storage", "showOpenFilePicker", "FileReader", "sqlite", "tauri", "invoke(", "vault", "export", "WebSocket", "OpenAI"]:
    check("CLOSED-" + forbidden, forbidden.lower() not in all_text.lower(), "forbidden capability token is absent")
payload = {"task":"LIFEOS-P3-090", "attempt":"rework/attempt-2", "runner":"independent_static_runner.py", "app":str(APP), "pass":sum(r["pass"] for r in results), "fail":sum(not r["pass"] for r in results), "results":results, "counts":{"P0":0,"P1":0,"P2":0,"Unknown":0,"NotImplemented":0}}
OUT.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
print(f"{payload['pass']} PASS / {payload['fail']} FAIL")
sys.exit(1 if payload["fail"] else 0)
